[fx] make sure args/kwargs are immutable

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45193 [fx] allow custom behavior for args, kwargs, and bool
* **#46121 [fx] make sure args/kwargs are immutable**

Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

zdevito committed:
[fx] make sure args/kwargs are immutable

Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

[ghstack-poisoned]

jamesr66a commented on torch/fx/immutable_collections.py:
```diff
@@ -0,0 +1,13 @@
+def _no_mutation(self, *args, **kwargs):
+    raise NotImplementedError(f"'{type(self).__name__}' object does not support mutation")
```
Can we tell the user to explicitly assign new `args` and `kwargs`? Just so it's obvious what do you if you've gotten here

jamesr66a reviewed (approved):
![image](https://user-images.githubusercontent.com/4685384/95639117-26d1ad80-0a4c-11eb-8e53-7d9c897542af.png)

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

zdevito committed:
Update on "[fx] make sure args/kwargs are immutable"


Otherwise, mutating them would make the uses/users lists inaccurate.
You can still mutate the node by assigning a new value to .args or .kwargs

Differential Revision: [D24232288](https://our.internmc.facebook.com/intern/diff/D24232288)

[ghstack-poisoned]

robieta commented:
This appears to have broken MyPy tests. (Observed in https://github.com/pytorch/pytorch/pull/46023) [CI Failure](https://fburl.com/k2v1s6jy)

mruberry commented:
Unlanding. This broke multiple builds, as @robieta says. Relevant log snippet:

```
test_run_mypy - TestTypeHints
test_type_hints.py

Traceback (most recent call last):
  File "test_type_hints.py", line 217, in test_run_mypy
    self.fail(f"mypy failed: {stdout} {stderr}")
AssertionError: mypy failed: torch/fx/node.py:36: error: Need type annotation for '_uses' (hint: "_uses: Dict[<type>, <type>] = ...")  [var-annotated]
torch/fx/node.py:37: error: Argument 1 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]"  [arg-type]
torch/fx/node.py:37: error: Argument 2 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Dict[str, Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]]"  [arg-type]
torch/fx/node.py:103: error: Argument 1 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Tuple[Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None], ...]"  [arg-type]
torch/fx/node.py:111: error: Argument 2 to "_update_args_kwargs" of "Node" has incompatible type "Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]"; expected "Dict[str, Union[Tuple[Any, ...], List[Any], Dict[str, Any], slice, Node, str, int, float, bool, dtype, Tensor, None]]"  [arg-type]
Found 5 errors in 1 file (checked 1099 source files)
```
